File: sonar-text-plugin/src/test/java/org/sonar/plugins/secrets/rules/GoogleApiKeyRuleTest.java
/*
 * SonarQube Text Plugin
 * Copyright (C) 2021-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.plugins.secrets.rules;

import java.util.List;
import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.groups.Tuple.tuple;
import static org.sonar.plugins.secrets.TestUtils.aNormalizedInputFile;

public class GoogleApiKeyRuleTest {
  private final GoogleApiKeyRule underTest = new GoogleApiKeyRule();

  @Test
  public void testRuleProperties() {
    assertThat(underTest.getRuleKey()).isEqualTo("S6334");
    assertThat(underTest.getMessage()).isEqualTo("Make sure this Google API Key is not disclosed.");
  }

  @Test
  public void testRuleRegexPositive() {
    List<Secret> secrets = underTest.findSecretsIn(aNormalizedInputFile("android:value=\"AIzaSyCis4NzxMw1aJyvUIrjGILjPkSdxrRfof4\""));

    assertThat(secrets)
      .extracting(Secret::getTextRange)
      .extracting("start.line", "start.lineOffset", "end.line", "end.lineOffset")
      .containsOnly(tuple(1, 15, 1, 54));
  }

  @Test
  public void testRuleRegexNegative() {
    List<Secret> secrets = underTest.findSecretsIn(aNormalizedInputFile("android:value=\"AIzaSyCis4NzxMw1aJyvUIrjGfof4\""));

    assertThat(secrets).isEmpty();
  }
}

File: sonar-text-dotnet/src/SonarLint.Secrets.DotNet.UnitTests/Rules/GoogleApiKeyRuleTest.cs
/*
 * SonarAnalyzer for Text
 * Copyright (C) 2021-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

// Ported from ...\sonar-secrets-plugin\src\test\java\com\sonarsource\secrets\rules\GoogleApiKeyRuleTest.java

using FluentAssertions;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using SonarLint.Secrets.DotNet.Rules;
using SonarLint.VisualStudio.Integration.UnitTests;
using System.Collections.Generic;
using System.Linq;
using static SonarLint.Secrets.DotNet.UnitTests.TestUtils;

namespace SonarLint.Secrets.DotNet.UnitTests.Rules
{
    [TestClass]
    public class GoogleApiKeyRuleTest
    {
        [TestMethod]
        public void MefCtor_CheckIsExported()
        {
            MefTestHelpers.CheckTypeCanBeImported<GoogleApiKeyRule, ISecretDetector>(null, null);
        }

        [TestMethod]
        public void TestRuleProperties()
        {
            var testSubject = new GoogleApiKeyRule();

            testSubject.RuleKey.Should().Be("secrets:S6334");
            testSubject.Name.Should().Be("Google API keys should not be disclosed");
            testSubject.Message.Should().Be("Make sure this Google API Key is not disclosed.");
        }

        [TestMethod]
        public void TestRuleRegexPositive()
        {
            var testSubject = new GoogleApiKeyRule();

            var input = "android:value=\"AIzaSyCis4NzxMw1aJyvUIrjGILjPkSdxrRfof4\"";

            var secrets = testSubject.Find("android:value=\"AIzaSyCis4NzxMw1aJyvUIrjGILjPkSdxrRfof4\"");

            secrets.Count().Should().Be(1);
            CheckExpectedSecretFound(input, "AIzaSyCis4NzxMw1aJyvUIrjGILjPkSdxrRfof4", secrets.First());
            CrossCheckWithJavaResult(input, new(1, 15, 1, 54), secrets.First());
        }

        [TestMethod]
        public void TestRuleRegexNegative()
        {
            var testSubject = new GoogleApiKeyRule();

            IEnumerable<ISecret> secrets = testSubject.Find("android:value=\"AIzaSyCis4NzxMw1aJyvUIrjGfof4\"");

            secrets.Should().BeEmpty();
        }
    }
}
